### DISCUSSION TITLE
How to clear vm state?

### DISCUSSION THREAD

**Discussion by kernelsoe** [sharkdp/numbat#625] [Q&A] [answered]
First of all, Thank you so much for this amazing numbat!!!

I'm walking through the source but still need some time to understand how the VM works.
I saw the `reset` function in web version but it's creating a new Numbat instance.

Is there a way to clear user defined functions, vars like `interpret` function?

**Comment by sharkdp** [comment]
I'm not sure I understand. Are you using Numbat as a library? Or as a CLI/Web tool?

**Reply by kernelsoe** [reply]
I'm using numbat as a rust library and interpreting user input strings from some text editor like:

```rs
ctx.interpret_with_settings(&mut settings, code, CodeSource::Text)
```

For e.g.
When the user first evaluates `let abc = 123` then delete the line, the abc is still available. I would like to clear the value `abc` from numbat instance.

Thanks in advance 🙏

**Reply by sharkdp** [reply, marked as answer]
I'm afraid this is not possible at the moment. What we do in the web-version is to replace the whole interpreter (https://github.com/sharkdp/numbat/blob/80b3f3444ff61f229a4a3ea956447bf44fa6f99d/numbat-wasm/www/index.js#L45). Something similar should also be possible in Rust.

You could cache/store the context after the prelude has been loaded, but before any user input has been evaluated. Whenever you want to reset, you go back to that cached state of the context. We do something similar in the unit tests, and it was a massive speed up for the test suite.
